stage1/init/kvm.go

```diff
@@ -105,21 +105,24 @@ func mountSharedVolumes(root string, p *stage1commontypes.Pod, ra *schema.Runtim
 			return errwrap.Wrap(fmt.Errorf("could not resolve symlink for source: %v", source), err)
 		} else if err := ensureDestinationExists(cleanedSource, absDestination); err != nil {
 			return errwrap.Wrap(fmt.Errorf("could not create destination mount point: %v", absDestination), err)
-		} else if err := doBindMount(cleanedSource, absDestination, readOnly); err != nil {
+		} else if err := doBindMount(cleanedSource, absDestination, readOnly, vol.Recursive); err != nil {
 			return errwrap.Wrap(fmt.Errorf("could not bind mount path %v (s: %v, d: %v)", m.Path, source, absDestination), err)
 		}
 	}
 	return nil
 }
 
-func doBindMount(source, destination string, readOnly bool) error {
-	if err := syscall.Mount(source, destination, "bind", syscall.MS_BIND, ""); err != nil {
-		return err
-	}
+func doBindMount(source, destination string, readOnly bool, recursive *bool) error {
+	var flags uintptr = syscall.MS_BIND | syscall.MS_REC
 	if readOnly {
-		return syscall.Mount(source, destination, "bind", syscall.MS_REMOUNT|syscall.MS_RDONLY|syscall.MS_BIND, "")
+		flags |= syscall.MS_RDONLY
 	}
-	return nil
+
+	// Enable recursive by default and remove it if explicitly requested
+	if recursive != nil && *recursive == false {
+		flags ^= syscall.MS_REC
+	}
+	return syscall.Mount(source, destination, "bind", flags, "")
```

Comment:
In case of read-only, there was previously two calls to syscall.Mount(). And with that patch, it is replaced by only one call to syscall.Mount(). I am not sure it will work, and it might depend on the kernel version. As far as I remember, the reason for that is that the kernel does not support bind-mounts with both readonly and recursive at the same time. That's why the function [bind_remount_recursive](https://github.com/systemd/systemd/blob/v231/src/basic/mount-util.c#L315) with its huge comment exists.

Implementing the 154 lines of code from [bind_remount_recursive](https://github.com/systemd/systemd/blob/v231/src/basic/mount-util.c#L315) in Golang here would be annoying, so I would prefer to be proven wrong.